src/asynchronous/tls.d

```diff
@@ -0,0 +1,745 @@
+/**
+ * TLS support.
+ *
+ * Copyright: © 2015-2016 Dragos Carp
+ * License: Boost Software License - Version 1.0
+ * Authors: Dragos Carp
+ */
+module asynchronous.tls;
+
+version (Have_botan):
+
+import std.stdio;
+import std.socket : SocketOSException;
+import std.typecons : Tuple, tuple, Nullable, Rebindable, rebindable;
+import botan.math.bigint.bigint;
+import botan.algo_base.symkey;
+import botan.tls.credentials_manager : TLSCredentialsManager;
+import botan.cert.x509.x509cert : X509Certificate;
+import botan.rng.auto_rng : RandomNumberGenerator, AutoSeededRNG;
+import botan.pubkey.pkcs8 : loadKey, PrivateKey;
+import botan.tls.channel;
+import botan.tls.server;
+import botan.tls.client;
+import asynchronous.transports;
+import asynchronous.futures : Waiter;
+import asynchronous.events : SslContext, EventLoop;
+import asynchronous.types : NotImplementedException;
+
+/** Supported TLS protocol versions */
+enum ProtocolVersion
+{
+    TLSv1_2,
+}
+
+/**
+ * Handshake failed.
+ */
+class HandshakeException : Exception
+{
+    this(string message = null, string file = __FILE__, size_t line = __LINE__,
+        Throwable next = null) @safe pure nothrow
+    {
+        import std.format;
+        if (message)
+        {
+            try {
+                super(format("SSL handshake failed: %s", message), file, line, next);
```

Comment:
I'm thinking not to use too many Exception types and just use somethink like TLSException for all TLS errors. This seems to be more common for D aswell (one exception type pro module). What do you think?